torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -572,20 +606,62 @@ void Reducer::prepare_for_backward(
 
 // A bucket with one or more dense tensors needs to be unflattened.
 void Reducer::finalize_bucket_dense(Bucket& bucket) {
-  for (auto& replica : bucket.replicas) {
+  for (size_t replica_index = 0; replica_index < bucket.replicas.size();
+       replica_index++) {
+    auto& replica = bucket.replicas[replica_index];
     for (size_t intra_bucket_index = 0;
          intra_bucket_index < replica.variables.size();
          intra_bucket_index++) {
       auto& variable = replica.variables[intra_bucket_index];
       const auto offset = replica.offsets[intra_bucket_index];
       const auto length = replica.lengths[intra_bucket_index];
+
+      // Determine if this param has been used globally or not.
+      //
+      // If the variable was used locally, it is also used globally and then
+      // we don't need to wait for the reduction. Otherwise we lazily wait for
+      // the reduction to complete, only when we see a variable that was unused
+      // locally. Then we end up delaying the synchronization point that
+      // local_used_work_->wait() implies. If we don't have any unused
+      // parameters at all, we can skip waiting for the work to complete
+      // altogether, and cause negligible performance overhead for models where
+      // all parameters are used. Such lazily waiting means minimizing
+      // performance impact for the big majority of models where all parameters
+      // are always used. Then we only pay the overhead cost if there is indeed
+      // a parameter that is locally unused, because we need to check if it's
+      // also globally unused.
+      size_t variable_index = bucket.variable_indices[intra_bucket_index];
+      // Note: global_unused might not be global yet. As we lazily wait for the
+      // reduction to complete, it becomes really global only if we get to the
+      // point as below where we wait for the reduction work and update
+      // global_unused with the real global consensus, i.e.
+      // local_used_maps_reduced_ is true.
+      bool global_unused =
+          (*local_used_maps_1_)[replica_index][variable_index].item<int>() == 0;
+      if (global_unused && !local_used_maps_reduced_) {
+        // Wait for local_used_maps reduction to complete.
+        local_used_work_->wait();
+        // At this point, local_used_maps_2_ has the global consensus. We swap
+        // their content.
+        std::swap(local_used_maps_1_, local_used_maps_2_);
```

Comment:
You can't do this yet here. NCCL collectives run asynchronously and the `wait()` call only synchronizes them with the current CUDA stream. This means only CUDA kernels will be sequenced after the NCCL kernel. Because you're accessing host memory here, it may still be running, and therefore get an incorrect result. You need to perform device synchronization to be certain you can access it.

Reply:
Yes, you are right. Host memory may not be ready yet at this point.